kod/object/active/holder/nomoveon/battler/player/user.kod

```diff
@@ -3685,6 +3695,123 @@ messages:
       return;
    }
 
+   UserGetFromContainer(items_list=$, numbers_list=$)
+   {
```

Comment:
The main issue here is that lots of this code was copied from UserGet().  (You can read about the "DRY principle" to learn why this is a problem).  So before I review this in detail, how about you make an attempt to share the common code between the two functions.

Reply:
I did think about this a bit. In terms of gameplay the two actions are fundamentally different.  In terms of coding things are murkier.  I'll start with gameplay and may ask about some coding options on Discord and then respond here or make the changes as needed.

The fundamental difference in terms of gameplay is that players will withdraw a stack of items from the chest once or twice per game session.  This is much like how a player will go to a vendor and buy a stack of reagents to be fully ready for a battle.  Here the user wants to manage their inventory and is planning for an activity.

Conversely picking items off the ground is an action the player will do 40-80+ times during a gameplay session if they are fighting monsters.  Here they are simply accumulating goods.  Each pickup is in the range of 2-5 items in small stacks.  I don't want the user to have to tell the game that yes, they want to pick up all 3 of those rainbow ferns.  In practice, most experienced players will pick up items from the ground with hotkeys.  For me this is 'G' for get menu, then down arrow to the first item I want, then spacebar to select, then arrows for more items and enter to pickup everything highlighted.  It would be a problem if the user had to mouse to the quantity tab for each pickup.

So in terms of player action there is a difference.  That said there is no need to turn on the LD_AMOUNTS flag for the get/look menu outside of GotObjectContents.  In that case the UserGet could have a default condition of number_items = $.  The only problem with that is that the UserGet routine does balloon in length a bit.  Now we will need to check if number_items <> $ in a bunch of spots before doing the oSplit function.

Another issue is one I struggled with greatly (I even had two working options coded).  This is somewhat fundamental.  The issue is: do we loop over the selected items in C and then UserGet each item one at a time in KOD or do we send over the list and then UserGetFromContainer the whole list in a loop in KOD.  In this case I opted for the second option because I wanted a success message when everything is picked up and I didn't want to add the two properties that would be required to do that while looping over the selection in C. (The process involved sending the length of the list as an argument and then incrementing a property for each loop and a second property for each successful pickup.  When the loop property equals the length argument if the success property also equals the length argument you can send the success message).

The result is that right now UserGetFromContainer loops over a list in KOD and UserGet loops over the list in C and picks up each element of the list one at a time.  If we want to combine these functions we are adding a bunch of checks for (Number_Items = $) and we are probably moving UserGetFromContainer away from the list method and towards looping in C.  If instead we want to change UserGet to loop in KOD that is do-able but seemed risky to me.  UserGet is pretty fundamental to the gameplay and any change of that scale would require some amount of additional testing.

That said, if this is the direction we want to go I think I can implement either of these options.

A third path exists as well: We can offload a lot of the safety checks and make these into sub-routines and then UserGet and UserGetFromContainer would both call these subroutines first and if that passes then we move onto the meat of getting the goods.  I think this would pull 20-30 lines out of each method.

Reply:
Attempt Made.  Sorry for long reply above.